### PR TITLE
Tensor tch functions

### PR DESCRIPTION
**Pull request by Athospd**
tch functions implemented:

- tch_log
- tch_log2
- tch_log10
- ttch_log1p
- tch_tan
- tch_tanh
- tch_sin
- tch_sinh
- tch_std
- tch_var
- tch_min
- tch_max
- tch_prod
- tch_logsumexp
- tch_mode <- **new**
- tch_median <- **new**
- tch_tril <- **new**
- tch_triu <- **new**
- tch_rsqrt <- **new**
- tch_sigmoid <- **new**
- tch_sign <- **new**
- tch_sqrt <- **new**
- tch_trunc <- **new**
- tch_lerp <- **new**

Methods:

- rsqrt_
- sigmoid_
- sin_
- tan_
- sqrt_
- tanh_
- trunc_
- abs_
- acos_
- asin_

tch functions that I failed to implement:

- ~tch_mode~
- ~tch_median~

Notes: 

- tch_logsumexp() seems to have some divergence between pytorch doc and c++ implementation. I'll discuss this in a issue.
- I kept the same structure of tch_mean() function for the major of these implementations, so they suffer of the same lack of other arguments code.

``` 
  // TODO handle other sum arguments.
  Rcpp::stop("Not yet implemented");
```

### CODE REVIEW

**Comment by dfalbel** [comment]
@Athospd I think it didn't work median and mode, because when you have a dim argument median will return 2 tensors: the medians for each row and the index of the median for each row. 

In this case I think we will need 2 functions: 1 to calculate the median (without any additional args) and another to calculate the median with and return both tensors to R. `tensor_qr_` has a reference implementation for this. Then we need to create some glue code on the R side.

**Comment by dfalbel** [comment]
@Athospd I created a reference implementation for `tch_mean` here: https://github.com/dfalbel/torch/blob/4bd94ae7db65f0dbf0308e1cca892eec359f13a0/src/tensor.cpp#L1037-L1049

We need to adapt other functions that have this kind of behavior to use this.

File: R/tensor-R6.R [hunk 1]
```diff
@@ -175,6 +175,11 @@
       `torch::Tensor`$dispatch(tensor_tanh_(self$pointer))
     },
 
+    tanh_ = function() {
+      tensor_tanh__(self$pointer)
+      invisible(NULL)
```
Comment (dfalbel) [inline R/tensor-R6.R]: @Athospd I think it's bets to use `invisible(self)` so users can chain method like `x$tanh_()$mean()`. Some old code are still using `invisible(NULL)`.

Comment (dfalbel) [reply R/tensor-R6.R]: I just added tests for all methods you created.

**Comment by Athospd** [comment]
@dfalbel I found out that the following functions has the tuple (dim, keepdim) args:

argmax
argmin
logsumexp
median
mode
norm
prod
std
sum
var
kthvalue
max
min

We have to adapt these using mean as reference, I got it right?

**Comment by dfalbel** [comment]
Exactly!

**Comment by dfalbel** [comment]
I will close this PR and merge #31 .

**Pull request closed without merging**